core/commands/dns.go

```diff
@@ -63,7 +63,7 @@ The resolver will give:
 			depth = namesys.DefaultDepthLimit
 		}
 		output, err := resolver.ResolveN(req.Context().Context, name, depth)
-		if err != nil {
+		if err != nil && (err != namesys.ErrResolveRecursion || depth > 1) {
```

Comment:
On Wed, Jun 10, 2015 at 01:26:04AM -0700, Juan Batiz-Benet wrote:
> > -		if err != nil {
> > +		if err != nil && (err != namesys.ErrResolveRecursion || depth > 1) {
>
> Wait, why is that an error? If `depth == 1` and that's the expected
> return, then it should not be an error, it should be the proper
> return.

You said “Resolver, I'd like you to completely resolve this path to
/ipfs/…, and I'll let you take one step”.  And the resolver says “I
took all the steps you let me take without trouble and got to <here>,
but I didn't make it to /ipfs/…”.  It's then up to the caller to
decide whether it counts as success (“we successfully took all the
requested steps! :)”) or failure (“we didn't make it to /ipfs/… :(”).
For recursive resolution, I expect the more important condition is
“did we make it to /ipfs/…?”.  But for non-recursive resolution, I
expect the more important condition is “was the step successfully
resolved to anything (/ipfs/… or otherwise)?”

As I said earlier [1], I think the current strategy of returning both
the resolved path and the ErrResolveRecursion error (“we successfully
took all our allotted steps to get to <path>, but it's not /ipfs/…”)
makes sense, but then callers need logic like this to decide if they
consider that an error.  But as I said in [1], I'd feel almost as good
about pushing this depth=1 special casing down into the resolve()
helper (and clients that didn't like that choice could explicitly
check for /ipfs/… paths in non-error, depth=1 cases).

[1]: https://github.com/ipfs/go-ipfs/pull/1351/files#r32077819